# Data Wrangling

This data is pulled from the CT MLS system from 7/28 - 7/29/21.  The listings are those which closed within the range of 0-365 days from the previous dates listed.  All citys available were pulled from CT.  Listings are limited to single family homes, condos, co-op.  In the steps below the raw data will be processed to remove null values, convert data into either categorical or numerical data, and check for outliers. 

In [2]:
import pandas as pd
import numpy as np

In [3]:
pwd

'C:\\Users\\kyu_k\\OneDrive\\Desktop\\ds\\Spring board DS Career track\\Capstonetwo\\Untitled Folder'

In [4]:
data_merged = pd.DataFrame()


In [5]:
#C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstone 2

In [6]:
for i in range(1,69):        ####read csvs into dataframes
    
    filename = 'C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (' + str(i) + ').csv'
    dataframe = pd.read_csv(filename)
    print(filename)
    
    if i == 1:
        data_merged = dataframe
        
        
    else:
        data_merged = data_merged.append(dataframe)
        
        
   

C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (1).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (2).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (3).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (4).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (5).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (6).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (7).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (8).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Spring board DS Career track/Capstonetwo/data/Agent Single Line (9).csv
C:/Users/kyu_k/OneDrive/Desktop/ds/Sp

In [7]:
data_merged.reset_index(inplace=True,drop = True)

In [8]:
print(data_merged.shape)

(58107, 21)


In [9]:
print(data_merged.head())

   #       MLS# Status Status Change Timestamp Prop Type  \
0  1  170323774   CLSD                09/13/20        SF   
1  1  170366202   CLSD                03/18/21        SF   
2  1  170378716   CLSD                05/14/21        SF   
3  1  170305413   CLSD                03/10/21        SF   
4  1  170365013   CLSD                04/22/21        SF   

               List/Closed Price           Address Unnamed: 7     City  Acres  \
0  LP: $3,750,000 CP: $3,750,000      213 Separate              Amenia  21.42   
1  LP: $7,950,000 CP: $6,800,000  292 Wood Duck Rd              Amenia   0.32   
2       LP: $150,000 CP: $97,500   64 Long Hill Rd             Andover   0.92   
3      LP: $119,900 CP: $100,000       560 Route 6             Andover   4.63   
4      LP: $170,000 CP: $150,000       7 Bailey Rd             Andover   1.49   

   ... Sq Ft Est Heated Above Grade Style or Rent Type Rooms Total  \
0  ...                        4,392           Colonial          11   
1  ...      

Features like garage/park and list/closed price we are going to have to split the feature into 2 new columns.

In [10]:
print(data_merged.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58107 entries, 0 to 58106
Data columns (total 21 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   #                             58107 non-null  int64  
 1   MLS#                          58107 non-null  object 
 2   Status                        58107 non-null  object 
 3   Status Change Timestamp       58107 non-null  object 
 4   Prop Type                     58107 non-null  object 
 5   List/Closed Price             58107 non-null  object 
 6   Address                       58107 non-null  object 
 7   Unnamed: 7                    58107 non-null  object 
 8   City                          58107 non-null  object 
 9   Acres                         46390 non-null  object 
 10  Sq Ft Total                   58107 non-null  object 
 11  Sq Ft Est Heated Above Grade  58090 non-null  object 
 12  Style or Rent Type            58106 non-null  object 
 13  R

Features like Acres need to be examined for null values.  Acres, Sqft, Bathrooms need to be converted into numeric values.

In [11]:
data_merged.describe()

#   Rooms Total    Beds Total    Year Built
count  58107.000000  58107.000000  58107.000000  58103.000000
mean    1692.818542      6.868381      3.149053   1964.649244
std     1466.930825      2.385229      1.023004     39.048374
min        1.000000      0.000000      0.000000      0.000000
25%        1.000000      5.000000      3.000000   1951.000000
50%     1501.000000      7.000000      3.000000   1969.000000
75%     3001.000000      8.000000      4.000000   1988.000000
max     4501.000000     99.000000     15.000000   2021.000000

------------------------------------------------------------------------------------------

# UNNAMED 7 

In [12]:
print(data_merged['Unnamed: 7'])

0         
1         
2         
3         
4         
        ..
58102     
58103     
58104     
58105     
58106     
Name: Unnamed: 7, Length: 58107, dtype: object


In [13]:
spaces=0
nulls=0
for i in data_merged['Unnamed: 7']:
    if i == ' ':
        spaces+=1
    elif i is None:
        nulls+=1
        

In [14]:
print(spaces,nulls)

58107 0


In [15]:
data_merged = data_merged.drop('Unnamed: 7',axis=1)  #### dropp column

Dropped Unnamed 7 feature as it contains no data

----------------------------------------------------------------------------------

# MLS Number 

MLS number is a unique identifier for each listing or observation

In [16]:

"""Checks for duplicate MLS IDs in Dataset
"""

'Checks for duplicate MLS IDs in Dataset\n'

In [17]:
data_merged['MLS#'].duplicated().describe() 

count     58107
unique        1
top       False
freq      58107
Name: MLS#, dtype: object

There are no duplicate values in MLS number

-----------------------------------------------------------------------

# Status

Status is whether listing is active or not.  Data was specifically pulled on closed listings

In [18]:
print(data_merged['Status'].unique())

['CLSD']



All the observations have closed as their status

***

# Status Date Change  

***

# Prop Type

Property type represents the type of ownership.  Data was limited to Single Family homes, condos, and co-operative

In [19]:
print(data_merged['Prop Type'].describe())

count     58107
unique        3
top          SF
freq      46390
Name: Prop Type, dtype: object


In [20]:

#prop_types = (data_merged['Prop Type'].unique())
#type_counts = []
#for i in prop_types:
#    count = [i,len(data_merged[data_merged['Prop Type']==i])]
#    type_counts.append(count)
#print(type_counts)


In [21]:
###[['SF', 46390], ['CO', 11468], ['CP', 249]]

In [22]:
data_merged['Prop Type'].value_counts()

SF    46390
CO    11468
CP      249
Name: Prop Type, dtype: int64

In [23]:
###double check totals match
data_merged['Prop Type'].value_counts().sum()

58107

All of the records are either Single Family, Condo, or Cooperative

***

# List/Closed Price 

List/Closed Price contains both the original listing price from the date the propert hit the market and the closed price when it was sold

In [24]:
print(data_merged['List/Closed Price'])


0        LP: $3,750,000 CP: $3,750,000
1        LP: $7,950,000 CP: $6,800,000
2             LP: $150,000 CP: $97,500
3            LP: $119,900 CP: $100,000
4            LP: $170,000 CP: $150,000
                     ...              
58102        LP: $675,000 CP: $705,000
58103    LP: $2,100,000 CP: $2,000,000
58104        LP: $889,900 CP: $930,000
58105    LP: $1,249,000 CP: $1,210,000
58106    LP: $2,300,000 CP: $2,500,000
Name: List/Closed Price, Length: 58107, dtype: object


In [25]:
print (data_merged.loc[0,'List/Closed Price'])

LP: $3,750,000 CP: $3,750,000


The list and closed price have to be separated into new columns and converted to numeric values 

In [26]:


###data_merged['list_price'] = 0 
###data_merged['close_price'] = 0
  

#for i in range(0,len(data_merged['List/Closed Price'])):
       
 
#    listLPCP = data_merged['List/Closed Price'][i].split(' ')
#    data_merged.loc[i,'close_price'] = listLPCP[3]
#    data_merged.loc[i,'list_price'] = listLPCP[1]
    
   

In [27]:
data_merged['List/Closed Price']=  data_merged['List/Closed Price'].apply(lambda x: x.split(' ') )

In [28]:
print(data_merged['List/Closed Price'].head())

0    [LP:, $3,750,000, CP:, $3,750,000]
1    [LP:, $7,950,000, CP:, $6,800,000]
2         [LP:, $150,000, CP:, $97,500]
3        [LP:, $119,900, CP:, $100,000]
4        [LP:, $170,000, CP:, $150,000]
Name: List/Closed Price, dtype: object


In [29]:
for i in data_merged['List/Closed Price']:                           ####checking position of items to confirm list price and closed price are in correct position
    if "L" not in i[0]:
        print('Check')
    elif "C" not in i[2]: 
         print('Check')

In [30]:
data_merged['list_price']=  data_merged['List/Closed Price'].apply(lambda x: x[1])    #####creates new columns

In [31]:
data_merged['close_price']=  data_merged['List/Closed Price'].apply(lambda x: x[3] )

In [32]:
data_merged = data_merged.drop('List/Closed Price',axis=1)  #### dropp column

### List Price

In [33]:
data_merged['list_price'].head()

0    $3,750,000
1    $7,950,000
2      $150,000
3      $119,900
4      $170,000
Name: list_price, dtype: object

Need to remove special characters

In [34]:
 #### remove special characters, converts to numeric

data_merged['list_price'] = data_merged['list_price'].str.replace('$','')
data_merged['list_price'] = data_merged['list_price'].str.replace(',','')
data_merged['list_price'] = data_merged['list_price'].str.replace('\xa0','')
data_merged['list_price'] = pd.to_numeric(data_merged['list_price']) 


Character \xa0 was the issue

### Close Price

In [35]:
data_merged['close_price'] = data_merged['close_price'].str.replace('$','')
data_merged['close_price'] = data_merged['close_price'].str.replace(',','')
data_merged['close_price'] = data_merged['close_price'].str.replace('\xa0','')
data_merged['close_price'] = pd.to_numeric(data_merged['close_price'])

In [36]:
data_merged['list_price'].describe().apply(lambda x: format(x, 'f')) ####apply to remove sci notation

count       58107.000000
mean       450050.699503
std        558295.500067
min          4900.000000
25%        210000.000000
50%        299900.000000
75%        469900.000000
max      16750000.000000
Name: list_price, dtype: object

In [37]:
data_merged['close_price'].describe().apply(lambda x: format(x, 'f'))  ####apply to remove sci notation

count       58107.000000
mean       447640.698504
std        538819.805630
min          7500.000000
25%        215000.000000
50%        306000.000000
75%        475000.000000
max      16500000.000000
Name: close_price, dtype: object

The following was used to process List/Closed price.  
1. List price was split from close price 
2. They are then put into their own columns
3. Special characters are removed both columns
4. The columns are converted into numeric

***

# Address 

House number and street name of property listing

In [38]:
data_merged['Address'].head()

0        213 Separate
1    292 Wood Duck Rd
2     64 Long Hill Rd
3         560 Route 6
4         7 Bailey Rd
Name: Address, dtype: object

In [39]:
data_merged['Address'].describe()

count           58107
unique          56877
top       15 Ridge Rd
freq                4
Name: Address, dtype: object

In [40]:
#df.apply(lambda x: func(x['col1'],x['col2']),axis=1)


In [41]:
#data_merged['Address']=  data_merged['Address'].apply(lambda x: x.split(' ',1) )  ###splits street name from house number

In [42]:
#data_merged['Address'].head()


-----------------------------------------------------------------------

# City

Name of the city where property listing is located

In [43]:
###Check for nulls 

In [44]:
data_merged['City'].value_counts().head(10)

Stamford         2048
Norwalk          1464
Waterbury        1324
Fairfield        1319
Bridgeport       1219
Danbury          1158
West Hartford    1126
Stratford        1077
Hamden           1007
Milford           975
Name: City, dtype: int64

In [45]:
data_merged['City'].isnull().value_counts()

False    58107
Name: City, dtype: int64

There are no nulls in the City column

_____________________________________________________________________________

# ACRES

Acres is the area of the land which the property sits

In [46]:
data_merged['Acres'].describe()  ###### check for nulls

count     46390.00
unique     1870.00
top           0.17
freq        942.00
Name: Acres, dtype: float64

In [47]:
(data_merged['Acres'].isnull()).value_counts()

False    46390
True     11717
Name: Acres, dtype: int64

In [48]:
((data_merged['Acres'].isnull()) == (data_merged['Prop Type']!='SF')).value_counts()   ### Checks if missing plot size are not single family homes

True    58107
dtype: int64

Nulls look like they are confined to none Single family dwellings

In [49]:
data_merged['Acres'].value_counts().head(10)

0.17    942
1.00    780
0.23    773
0.11    727
0.14    696
0.18    631
0.29    573
0.46    570
0.28    555
0.00    547
Name: Acres, dtype: int64

In [50]:
data_merged['Acres'].describe()

count     46390.00
unique     1870.00
top           0.17
freq        942.00
Name: Acres, dtype: float64

In [51]:
data_merged['Acres'].dtypes

dtype('O')

In [52]:
data_merged['Acres'] = data_merged['Acres'].apply(lambda x: str(x).replace(',','') if (',' in str(x)) else x )

The Acres column is mixed numeric objects and string objects

In [53]:
#for i in data_merged['Acres']:
#    if type(i) != float:
#            
#        print(i)
#        print(float(i))
#        
#        print(type(i))
#        print(i.isnumeric())
#       
        


In [54]:
data_merged['Acres'] = data_merged['Acres'].apply(lambda x: float(x))  ### Covert all obseravtions to float objects

In [55]:
data_merged['Acres'].describe() 

count    46390.000000
mean         1.992514
std         89.135904
min          0.000000
25%          0.240000
50%          0.480000
75%          1.140000
max      12911.000000
Name: Acres, dtype: float64

The ramianing nulls in the Acre column will be converted to 0's.  The already existing 0's should be examined

In [56]:
len(data_merged.loc[data_merged['Acres']==0 ])

673

There are potentially mislabelled Prop type if Single family dwellings have 0 acres recorded

In [57]:
len(data_merged.loc[ ((data_merged['Acres']== 0) 
                & (data_merged['Prop Type'] =='SF')) 
                ])

673

In [58]:
data_merged.loc[ (data_merged['Acres']== 0)]['Address']

470               10 Owl Ct, Unit #10
907          3 Lakeview Rise, Unit #3
913       16 Twin Oaks Trl, Unit #134
915                  10 Woodland Path
978                   501 Berlin Tpke
                     ...             
57544              lot 49 Heritage Hl
57550              Lot 15 Heritage Hl
57565        15 Heritage Ct, Unit #15
57571    32 Constitution Dr, Unit #32
57965            25 197 Rte, Unit #41
Name: Address, Length: 673, dtype: object

In [59]:
test =''
for i in data_merged['Address'].loc[ (data_merged['Acres']== 0)]:
          
        test = test + i
        

mapping = [ (',', ''), ('1', ''), ('2', ''), ('3', ''), ('4', ''), ('5', ''), ('6', ''), ('7', ''), ('8', ''), ('9', ''), ('0', '') ]
for a, b in mapping:
    test = test.replace(a, b)
test = test.split(' ')
test = pd.DataFrame(test)

In [60]:
test.value_counts()[test.value_counts()>5]

Unit          360
#             328
Rd            186
Dr             87
St             73
Ln             66
Ct             43
Cir            37
Way            35
Old            22
Main           21
Ave            19
Trl            16
Brook          15
Hill           14
Colony         13
Woodland       13
Park           13
R              12
Ter            12
Hunters        12
N.             12
Kelly          11
S.             11
Village        11
               11
Heritage        9
Cambridge       9
#A              9
Copper          9
Windermere      9
Ridge           9
Laurel          8
Colchester      8
Island          8
Tpke            8
Leetes          7
Prospect        7
Carson          7
Cmns            7
Allentown       7
And             6
Meadow          6
Whitewood       6
Circle          6
Pond            6
Shore           6
North           6
dtype: int64

Potentially 360 of these obeservations are mislabeled having 0 acres and unit in address suggesting condo or leased land
Should these be recoded.  They will be dropped for now

In [61]:
data_merged = data_merged[data_merged['Acres'] != 0]

In [62]:
data_merged['Acres'].isnull().value_counts()

False    45717
True     11717
Name: Acres, dtype: int64

In [63]:
data_merged['Acres'] = data_merged['Acres'].apply(lambda x: 0 if (np.isnan(x)) else x)   ### Turn Nulls into 0 

In [64]:
data_merged['Acres'].describe()

count    57434.000000
mean         1.609373
std         80.112502
min          0.000000
25%          0.130000
50%          0.340000
75%          1.000000
max      12911.000000
Name: Acres, dtype: float64

The Acres feature was processed as follows:
1. Was checked for null values.  Accounted for 12k observations that corresponded to condo/co-op properties
2. Commas are removed from numerical strings, strings are then converted to float
3. There are 673 obeservations with 0 acres reported.  362 maybe mislabeled condos.  All 673 dropped
4. Null values are recoded to 0

***

# SQFT TOTAL

Total square feet of living space in the listing building

In [65]:
data_merged['Sq Ft Total']

0        4,392
1        5,169
2        1,800
3        1,724
4        1,138
         ...  
58102    2,632
58103    2,467
58104    2,974
58105    4,047
58106    3,083
Name: Sq Ft Total, Length: 57434, dtype: object

In [66]:
data_merged['Sq Ft Total'] = data_merged['Sq Ft Total'].str.replace(',','')   #### Remove commas

In [67]:
data_merged['Sq Ft Total'] = pd.to_numeric(data_merged['Sq Ft Total'])        #### Convert to numeric

In [68]:
data_merged['Sq Ft Total'].head()

0    4392
1    5169
2    1800
3    1724
4    1138
Name: Sq Ft Total, dtype: int64

In [69]:
data_merged['Sq Ft Total'].describe()

count    57434.000000
mean      2152.290925
std       1449.656289
min          0.000000
25%       1302.000000
50%       1793.000000
75%       2565.000000
max      99999.000000
Name: Sq Ft Total, dtype: float64

In [70]:
data_merged[data_merged['Sq Ft Total']==0]

#       MLS# Status Status Change Timestamp Prop Type  \
42942  1001  E10196431   CLSD                10/19/20        CO   

                         Address        City  Acres  Sq Ft Total  \
42942  4 Northwest St, Unit #B44  Stonington    0.0            0   

      Sq Ft Est Heated Above Grade  ... Rooms Total  Beds Total  Bathrooms  \
42942                          NaN  ...           0           0        NaN   

      Garage/Park Year Built    DOM    Listing Agent List Office list_price  \
42942  1 Car/Open     1980.0  1,288  Kathy Michalove    SEABRD00      26000   

       close_price  
42942        23000  

[1 rows x 21 columns]

In [71]:
data_merged.drop([42942],inplace=True)      ### Remove 0 Values

In [72]:
data_merged[data_merged['Sq Ft Total']==0]


Empty DataFrame
Columns: [#, MLS#, Status, Status Change Timestamp, Prop Type, Address, City, Acres, Sq Ft Total, Sq Ft Est Heated Above Grade, Style or Rent Type, Rooms Total, Beds Total, Bathrooms, Garage/Park, Year Built, DOM, Listing Agent, List Office, list_price, close_price]
Index: []

[0 rows x 21 columns]

In [73]:
data_merged[data_merged['Sq Ft Total']==99999]

#       MLS# Status Status Change Timestamp Prop Type  \
13781     1  170369089   CLSD                03/30/21        CP   
42943  1001  170192538   CLSD                10/19/20        CO   

                         Address        City  Acres  Sq Ft Total  \
13781              5 Putnam Park   Greenwich    0.0        99999   
42943  4 Northwest St, Unit #B43  Stonington    0.0        99999   

      Sq Ft Est Heated Above Grade  ... Rooms Total  Beds Total  Bathrooms  \
13781                       99,999  ...           5           2     1 Full   
42943                       99,999  ...           1           0     2 Full   

                                   Garage/Park Year Built  DOM  \
13781   Car/Detached Garage, On Street Parking     1950.0   20   
42943                   Car/Unassigned Parking     1983.0  484   

         Listing Agent List Office list_price  close_price  
13781    Karen Oztemel     BHHS135     399000       399000  
42943  Kathy Michalove    SEABRD00      27000        27000  

[2 rows x 21 columns]

In [74]:
data_merged.drop([13781],inplace=True)                    #### Remove sqft == 99999
data_merged.drop([42943],inplace=True)

In [75]:
data_merged[data_merged['Sq Ft Total']==99999]

Empty DataFrame
Columns: [#, MLS#, Status, Status Change Timestamp, Prop Type, Address, City, Acres, Sq Ft Total, Sq Ft Est Heated Above Grade, Style or Rent Type, Rooms Total, Beds Total, Bathrooms, Garage/Park, Year Built, DOM, Listing Agent, List Office, list_price, close_price]
Index: []

[0 rows x 21 columns]

In [76]:
data_merged['Sq Ft Total'].describe()

count    57431.000000
mean      2148.920949
std       1329.701949
min         50.000000
25%       1302.000000
50%       1793.000000
75%       2565.000000
max      61184.000000
Name: Sq Ft Total, dtype: float64

Total sqft was processed as follows:
1. Special characters removed then string converted into numeric values
2. 1 observation removed with 0 or 99999 reported sqft 

----------------------------------------------------------------------------------

# Sq Ft Est Heated Above Grade 

Sqft heated above grade represents the Sqft of living space that doesn't include space from a finished basement.

In [77]:
data_merged['Sq Ft Est Heated Above Grade'].describe()

count     57415
unique     5371
top       1,008
freq        413
Name: Sq Ft Est Heated Above Grade, dtype: object

In [78]:
data_merged['Sq Ft Est Heated Above Grade'].head()

0    4,392
1    5,169
2    1,800
3    1,724
4    1,138
Name: Sq Ft Est Heated Above Grade, dtype: object

In [79]:
data_merged['Sq Ft Est Heated Above Grade'] = data_merged['Sq Ft Est Heated Above Grade'].str.replace(',','')

In [80]:
data_merged['Sq Ft Est Heated Above Grade'] = pd.to_numeric(data_merged['Sq Ft Est Heated Above Grade'])    

In [81]:
data_merged['Sq Ft Est Heated Above Grade'].describe()

count    57415.000000
mean      1943.185614
std       1169.549355
min         50.000000
25%       1209.000000
50%       1592.000000
75%       2323.000000
max      61184.000000
Name: Sq Ft Est Heated Above Grade, dtype: float64

In [82]:
len(data_merged[data_merged['Sq Ft Est Heated Above Grade']<=350])

13

In [83]:
data_merged['Sq Ft Est Heated Above Grade'].isnull().value_counts()

False    57415
True        16
Name: Sq Ft Est Heated Above Grade, dtype: int64

In [84]:
##data_merged[data_merged['Sq Ft Est Heated Above Grade'].isnull()]   

Suppressed output, null values have reported SQFT total so they wont be dropped.  They will be changed to the value in Sqft total

In [85]:
data_merged['Sq Ft Est Heated Above Grade'] = data_merged['Sq Ft Est Heated Above Grade'].apply(lambda x: 0 if (np.isnan(x)) else x)

In [86]:
len(data_merged[data_merged['Sq Ft Est Heated Above Grade']==0])

16

In [87]:
data_merged['Sq Ft Est Heated Above Grade'].describe()

count    57431.000000
mean      1942.644251
std       1169.836016
min          0.000000
25%       1209.000000
50%       1591.000000
75%       2322.500000
max      61184.000000
Name: Sq Ft Est Heated Above Grade, dtype: float64

In [88]:
data_merged.loc[data_merged['Sq Ft Est Heated Above Grade']==0, 'Sq Ft Est Heated Above Grade']=data_merged.loc[data_merged['Sq Ft Est Heated Above Grade']==0, 'Sq Ft Total']

All the 0 values are filled with the corresponding Sqft total as a substitute

In [89]:
len(data_merged[data_merged['Sq Ft Est Heated Above Grade']==0])

0

In [90]:
data_merged['Sq Ft Est Heated Above Grade'].describe()

count    57431.000000
mean      1943.349097
std       1169.888998
min         50.000000
25%       1209.000000
50%       1592.000000
75%       2323.000000
max      61184.000000
Name: Sq Ft Est Heated Above Grade, dtype: float64

Sqft above grade was processed as follows:
1. Special characters have been removed, and converted to numeric entries
2. There are 13 entries that have less than 350 sqft but greater than 0
3. 16 entries that were null were converted to zeros.  They still had reported Sqft totals so they were filled with Sqft Totals

--------------------------------------------------------------------------

# Style or Rent Type 

In [91]:
data_merged['Style or Rent Type'].head() 

0        Colonial
1    Contemporary
2        Colonial
3           Ranch
4        Bungalow
Name: Style or Rent Type, dtype: object

In [92]:
data_merged['Style or Rent Type'].describe() 

count        57430
unique         254
top       Colonial
freq         16157
Name: Style or Rent Type, dtype: object

In [93]:
data_merged['Style or Rent Type'].value_counts().head(10) 

Colonial                  16157
Ranch                     12631
Cape Cod                   7365
Townhouse                  6233
Raised Ranch               3341
Split Level                1511
Contemporary               1260
Colonial, Contemporary      657
Other                       493
Colonial, Antique           453
Name: Style or Rent Type, dtype: int64

In [94]:
len(data_merged[data_merged['Style or Rent Type'].isnull()])

1

In [95]:
data_merged.loc[data_merged['Style or Rent Type'].isnull(),['Style or Rent Type']] = 'Other'

1 null observation was converted to Other

In [96]:
data_merged['Style or Rent Type'].str.isnumeric().value_counts()

False    57431
Name: Style or Rent Type, dtype: int64

In [97]:
data_merged['Style or Rent Type'].str.isalpha().value_counts()

True     38145
False    19286
Name: Style or Rent Type, dtype: int64

In [98]:
data_merged['Style or Rent Type'].str.isdigit().value_counts()

False    57431
Name: Style or Rent Type, dtype: int64

There are no Entrys which contain numeric characters, 20k obsevations contain special characters likely commas and spaces

In [99]:
data_merged['Style or Rent Type'] = data_merged['Style or Rent Type'].apply(lambda x: x.split(', ') if(',' in x) else x)

In [100]:
data_merged['Style or Rent Type'] = data_merged['Style or Rent Type'].apply(lambda x: x[0] if(type(x) is list) else x)

Properties with a more than one style are split and is replaced by the first entry

In [101]:
data_merged['Style or Rent Type'].value_counts().head(30) 

Colonial                  18484
Ranch                     13617
Cape Cod                   8569
Townhouse                  6438
Raised Ranch               3440
Contemporary               1897
Split Level                1543
Other                       494
Bungalow                    433
Single Family Detached      364
Apartment                   320
Antique                     298
Cottage                     228
Farm House                  179
Half Duplex                 167
Mid Rise                    147
High Rise                   141
Victorian                   123
Tudor                        78
Hi-Ranch                     73
Barn                         51
Georgian Colonial            43
Log                          42
Modern                       39
Row House                    37
Saltbox                      32
Penthouse                    28
Mobile Home                  22
European                     22
Studio Apt                   20
Name: Style or Rent Type, dtype: int64

Style is processed as follows:
1. The 1 null entry is replaced by 'other'
2. If multiple styles are used they are replaced by the first style in observation

---------------------------------------------------------------------

# Rooms Total 

Rooms total represents the total rooms in a property.  Included will be dining rooms, bedrooms, family rooms etc

In [102]:
data_merged['Rooms Total'].head()

0    11
1     7
2     7
3     6
4     6
Name: Rooms Total, dtype: int64

In [103]:
data_merged['Rooms Total'].describe()

count    57431.000000
mean         6.886577
std          2.386578
min          1.000000
25%          5.000000
50%          7.000000
75%          8.000000
max         99.000000
Name: Rooms Total, dtype: float64

In [104]:
len(data_merged[data_merged['Rooms Total']==99])

2

In [105]:
data_merged.drop([14307],inplace=True)                    #### Remove sqft == 99
data_merged.drop([14308],inplace=True)

In [106]:
len(data_merged[data_merged['Rooms Total']==99])

0

2 observations are dropped with Rooms total ==0.  They had missing entries across the features

In [107]:
len(data_merged[data_merged['Rooms Total']==90])

1

In [108]:
data_merged.drop([5741],inplace=True)

In [109]:
len(data_merged[data_merged['Rooms Total']==90])

0

1 Observation with 90 rooms dropped due to it being land and not a house

In [110]:
data_merged['Rooms Total'].isnull().value_counts()

False    57428
Name: Rooms Total, dtype: int64

In [111]:
data_merged['Rooms Total'].describe()

count    57428.000000
mean         6.881922
std          2.297877
min          1.000000
25%          5.000000
50%          7.000000
75%          8.000000
max         66.000000
Name: Rooms Total, dtype: float64

Total rooms was processed as follows:
1. All values in series are already numeric
2. 3 Observations dropped which had issues
3. No null values in the series

***

#   Beds Total

Beds Total represents the number of bedrooms in the house.  Usually characterized by a room with a door and closet.

In [112]:
data_merged['Beds Total'].describe()

count    57428.000000
mean         3.157223
std          1.023021
min          0.000000
25%          3.000000
50%          3.000000
75%          4.000000
max         15.000000
Name: Beds Total, dtype: float64

In [113]:
len(data_merged[data_merged['Beds Total']==0])

88

In [114]:
##data_merged[data_merged['Beds Total']==0]

In [115]:
data_merged['Beds Total'].isnull().value_counts()

False    57428
Name: Beds Total, dtype: int64

In [116]:
data_merged['Beds Total'].value_counts()

3     24530
4     14581
2     11471
5      3587
1      2164
6       794
7       156
0        88
8        35
9        12
11        5
12        2
10        2
15        1
Name: Beds Total, dtype: int64

Beds Total were processed as follows:
1. Values in series already numeric
2. Values in series seem reasonable 0 values can be studios
3. No modifications needed


--------------------------------------------------------------------------------------------------------------------

# Bathrooms 

Bathrooms represents the number of bathrooms in a property. They can either be half or full indicating the presence of shower/bathtub.

In [117]:
data_merged['Bathrooms'].describe()

count               57412
unique                 52
top       2 Full & 1 Half
freq                12880
Name: Bathrooms, dtype: object

In [118]:
data_merged['Bathrooms'].value_counts().head(10)

2 Full & 1 Half    12880
2 Full             12360
1 Full & 1 Half    10044
1 Full              9793
3 Full & 1 Half     3895
3 Full              3363
4 Full & 1 Half     1171
2 Full & 2 Half      716
4 Full               526
3 Full & 2 Half      519
Name: Bathrooms, dtype: int64

In [119]:
#  data_merged['Bathrooms'] = data_merged['Bathrooms'].apply(lambda x: x.split('&'))


AttributeError   Traceback (most recent call last)
<ipython-input-461-a8c3442477bf> in <module>
----> 1 data_merged['Bathrooms'] = data_merged['Bathrooms'].apply(lambda x: x.split('&'))

~\anaconda3\lib\site-packages\pandas\core\series.py in apply(self, func, convert_dtype, args, **kwds)
   4198             else:
   4199                 values = self.astype(object)._values
-> 4200                 mapped = lib.map_infer(values, f, convert=convert_dtype)
   4201 
   4202         if len(mapped) and isinstance(mapped[0], Series):

pandas\_libs\lib.pyx in pandas._libs.lib.map_infer()

<ipython-input-461-a8c3442477bf> in <lambda>(x)
----> 1 data_merged['Bathrooms'] = data_merged['Bathrooms'].apply(lambda x: x.split('&'))

AttributeError: 'float' object has no attribute 'split'

In [120]:
type(data_merged['Bathrooms'][1])     ####need to identify not str

str

In [121]:
test =data_merged[['MLS#','Bathrooms']]
issues = 0
issues_mls=[]
for i in range(0, len(test)-1):
    
    if type(test.iloc[i]['Bathrooms']) != str:
        issues +=1
        print(test.iloc[i])
        issues_mls.append(test.iloc[i]['MLS#'])

MLS#         170318789
Bathrooms          NaN
Name: 14587, dtype: object
MLS#         170296008
Bathrooms          NaN
Name: 14590, dtype: object
MLS#         170340976
Bathrooms          NaN
Name: 14591, dtype: object
MLS#         170394127
Bathrooms          NaN
Name: 14592, dtype: object
MLS#         170324757
Bathrooms          NaN
Name: 14593, dtype: object
MLS#         170262389
Bathrooms          NaN
Name: 14594, dtype: object
MLS#         170273381
Bathrooms          NaN
Name: 20279, dtype: object
MLS#         170257507
Bathrooms          NaN
Name: 29945, dtype: object
MLS#         170386295
Bathrooms          NaN
Name: 30479, dtype: object
MLS#         170350695
Bathrooms          NaN
Name: 34705, dtype: object
MLS#         170355886
Bathrooms          NaN
Name: 37384, dtype: object
MLS#         170384384
Bathrooms          NaN
Name: 42080, dtype: object
MLS#         170382922
Bathrooms          NaN
Name: 42945, dtype: object
MLS#         170399407 CO
Bathrooms             NaN

In [122]:
print(issues)

16


In [123]:
for i in issues_mls:                                   ####Checking what observations with missing baths looks like
    print(data_merged[data_merged['MLS#']==i])

         #       MLS# Status Status Change Timestamp Prop Type  \
14587  501  170318789   CLSD                08/10/20        CO   

                              Address      City  Acres  Sq Ft Total  \
14587  379 Whitfield St, Unit #CS1036  Guilford    0.0          578   

       Sq Ft Est Heated Above Grade  ... Rooms Total  Beds Total  Bathrooms  \
14587                         578.0  ...           1           0        NaN   

                                Garage/Park Year Built  DOM  \
14587   Car/Unassigned Parking, Parking Lot     2012.0    9   

                  Listing Agent List Office list_price  close_price  
14587  Deborah Citelli-Robinson      PEAR51      13500        12000  

[1 rows x 21 columns]
         #       MLS# Status Status Change Timestamp Prop Type  \
14590  501  170296008   CLSD                08/20/20        CO   

                              Address      City  Acres  Sq Ft Total  \
14590  379 Whitfield St, Unit #CS2636  Guilford    0.0          578   


In [124]:
data_merged=data_merged.dropna(subset = ['Bathrooms'])     #### Subset picks column for NA

16 observations dropped for having null values for bathrooms

In [125]:
 data_merged['Bathrooms'].apply(lambda x: 'Full' in x).value_counts()      ###### Count countaining full for QA

True     57402
False       10
Name: Bathrooms, dtype: int64

In [126]:
 data_merged['Bathrooms'].apply(lambda x: 'Half' in x).value_counts()        ###### Count countaining Half for QA

True     31220
False    26192
Name: Bathrooms, dtype: int64

In [127]:
data_merged['Bathrooms'] = data_merged['Bathrooms'].apply(lambda x: x.split('&'))

In [128]:
data_merged['Bathrooms'] 

0        [3 Full ,  1 Half]
1        [5 Full ,  1 Half]
2                  [1 Full]
3                  [1 Full]
4                  [1 Full]
                ...        
58102    [3 Full ,  1 Half]
58103    [2 Full ,  1 Half]
58104    [3 Full ,  1 Half]
58105    [3 Full ,  1 Half]
58106    [2 Full ,  1 Half]
Name: Bathrooms, Length: 57412, dtype: object

The series is split on '&' in order to extract both full and half baths

### Full Bath

In [129]:
data_merged['Full Bath'] = data_merged['Bathrooms'].apply(lambda x: x[0] if ('Full' in x[0]) else '0')   

In [130]:
data_merged['Full Bath'].head()

0    3 Full 
1    5 Full 
2     1 Full
3     1 Full
4     1 Full
Name: Full Bath, dtype: object

In [131]:
data_merged['Full Bath'].value_counts()

2 Full      13617
2 Full      12360
1 Full      10398
1 Full       9793
3 Full       4450
3 Full       3363
4 Full       1576
5 Full        692
4 Full        526
6 Full        299
7 Full        117
5 Full        117
8 Full         47
6 Full         27
0              10
9 Full          9
7 Full          4
12 Full         2
99 Full         1
20 Full         1
11 Full         1
9 Full          1
10 Full         1
Name: Full Bath, dtype: int64

In [132]:
data_merged['Full Bath'] = data_merged['Full Bath'].apply(lambda x: x.split(' ',1))

In [133]:
data_merged['Full Bath'].head()

0    [3, Full ]
1    [5, Full ]
2     [1, Full]
3     [1, Full]
4     [1, Full]
Name: Full Bath, dtype: object

In [134]:
data_merged['Full Bath'] = data_merged['Full Bath'].apply(lambda x: x[0])

In [135]:
data_merged['Full Bath'].value_counts()

2     25977
1     20191
3      7813
4      2102
5       809
6       326
7       121
8        47
0        10
9        10
12        2
20        1
11        1
99        1
10        1
Name: Full Bath, dtype: int64

In [136]:
data_merged = (data_merged[data_merged['Full Bath']!='99']) 

In [137]:
data_merged['Full Bath'] = pd.to_numeric(data_merged['Full Bath'])  

In [138]:
data_merged['Full Bath'].describe()

count    57411.000000
mean         1.939890
std          0.975377
min          0.000000
25%          1.000000
50%          2.000000
75%          2.000000
max         20.000000
Name: Full Bath, dtype: float64


Full Baths seems to be in order. Now to finish Half Baths 



### Half Baths

In [139]:
data_merged['Half Bath'] = data_merged['Bathrooms'].apply(lambda x: x[1] if (len(x)==2) else '0')

In [140]:
data_merged['Half Bath'].value_counts()

 1 Half    28592
0          26192
 2 Half     2393
 3 Half      207
 4 Half       25
 5 Half        1
 8 Half        1
Name: Half Bath, dtype: int64

In [141]:
data_merged['Half Bath'].head()

0     1 Half
1     1 Half
2          0
3          0
4          0
Name: Half Bath, dtype: object

In [142]:
data_merged['Half Bath'] = data_merged['Half Bath'].str.replace(' ','') 

In [143]:
data_merged['Half Bath'].head()

0    1Half
1    1Half
2        0
3        0
4        0
Name: Half Bath, dtype: object

In [144]:
data_merged['Half Bath'] = data_merged['Half Bath'].apply(lambda x: x.split('Half'))

In [145]:
data_merged['Half Bath'].head()

0    [1, ]
1    [1, ]
2      [0]
3      [0]
4      [0]
Name: Half Bath, dtype: object

In [146]:
data_merged['Half Bath'] = data_merged['Half Bath'].apply(lambda x: x[0])

In [147]:
data_merged['Half Bath'].value_counts()

1    28592
0    26192
2     2393
3      207
4       25
8        1
5        1
Name: Half Bath, dtype: int64

There are too many zero entries, perhaps there is an issue.  Upon Checking this matches the check before making changes to the bathroom column.

In [148]:
data_merged['Half Bath'] = pd.to_numeric(data_merged['Half Bath'])  

In [149]:
data_merged['Half Bath'].describe()

count    57411.000000
mean         0.594172
std          0.593872
min          0.000000
25%          0.000000
50%          1.000000
75%          1.000000
max          8.000000
Name: Half Bath, dtype: float64

In [150]:
####split on full or half, include else statement to "" or 0 for nothing #######

In [151]:
data_merged = data_merged.drop('Bathrooms',axis=1)  #### dropp column

Bathrooms were processed as follows:
1. 16 observations were dropped due to null bathrooms
2. The series was split into 2 columns for half and full baths
3. All non-numeric characters were removed
4. The series was converted to numeric values
5. All values seem reasonable. 1 dropped for 99 full bathroms


-------------------------------------------------------------------------------------

# Garage / park

Garage/park represents the number of space available for cars and their format.  Format can be attached garage, detached garage, off street park, carport, etc.

In [152]:
data_merged['Garage/Park'].head()

0       2 Car/Attached Garage
1       2 Car/Attached Garage
2    2 Car/Off Street Parking
3      Car/Off Street Parking
4       2 Car/Detached Garage
Name: Garage/Park, dtype: object

In [153]:
data_merged['Garage/Park'].value_counts().head(20)

2 Car/Attached Garage                               12918
1 Car/Attached Garage                                5814
1 Car/Detached Garage                                3202
 Car/Off Street Parking                              3031
2 Car/Detached Garage                                3028
2 Car/Under House Garage                             2394
1 Car/Under House Garage                             2391
3 Car/Attached Garage                                2191
 Car/Paved                                           1811
 Car/Paved, Off Street Parking                       1705
2 Car/Attached Garage, Paved                         1062
 Car/Assigned Parking                                 835
2 Car/Attached Garage, Under House Garage             718
 Car/Other                                            651
 Car/Unpaved                                          395
1 Car/Attached Garage, Under House Garage             388
3 Car/Detached Garage                                 350
1 Car/Attached

In [154]:
 data_merged['Garage/Park'].apply(lambda x: 'Garage' in x).value_counts()       ###counts entries with 'garage' for test below

True     43134
False    14277
Name: Garage/Park, dtype: int64

43k observations have potentially have garage. 14k potentially have no garage

In [155]:
data_merged['Garage/Park'] = data_merged['Garage/Park'].apply(lambda x: x.split('/',1))

In [156]:
data_merged['Garage/Park'].head()

0       [2 Car, Attached Garage]
1       [2 Car, Attached Garage]
2    [2 Car, Off Street Parking]
3     [ Car, Off Street Parking]
4       [2 Car, Detached Garage]
Name: Garage/Park, dtype: object

In [157]:
data_merged[data_merged['Garage/Park'].apply(lambda x: len(x)<2)]

#       MLS# Status Status Change Timestamp Prop Type  \
28528     1  N10197464   CLSD                05/09/21        CO   
32142  3501    L152330   CLSD                08/19/20        SF   

                            Address       City  Acres  Sq Ft Total  \
28528  40 Harbour Close, Unit #B-14  New Haven    0.0         1500   
32142          419 Bald Mountain Rd    Norfolk  518.3         7100   

       Sq Ft Est Heated Above Grade  ... Beds Total         Garage/Park  \
28528                        1500.0  ...          2  [Assigned Parking]   
32142                        7100.0  ...          9             [2 Car]   

       Year Built    DOM    Listing Agent List Office list_price close_price  \
28528      1983.0     43  Rick Chamberlin      REAT30     279900      265000   
32142      1925.0  2,089   THOMAS MCGOWAN      HARN42    3999000     4015000   

       Full Bath  Half Bath  
28528          2          1  
32142          7          0  

[2 rows x 22 columns]

In [158]:
### df.at[index, 'col_name'] = x

In [159]:
data_merged.at[32142,'Garage/Park'] = ['2 Car', 'Garage']

In [160]:
data_merged.at[28528,'Garage/Park'] = ['Assigned Parking','Parking Lot']

In [161]:
len(data_merged[data_merged['Garage/Park'].apply(lambda x: len(x)<2)])

0

In [162]:
data_merged[data_merged['Garage/Park'].apply(lambda x: len(x)>2)]

Empty DataFrame
Columns: [#, MLS#, Status, Status Change Timestamp, Prop Type, Address, City, Acres, Sq Ft Total, Sq Ft Est Heated Above Grade, Style or Rent Type, Rooms Total, Beds Total, Garage/Park, Year Built, DOM, Listing Agent, List Office, list_price, close_price, Full Bath, Half Bath]
Index: []

[0 rows x 22 columns]

Corrected 2 observations by looking them up on Zillow

In [163]:
data_merged['Garage'] = data_merged['Garage/Park'].apply(lambda x: x[0] if ('Garage' in x[1]) else '0')   ####Uses count if garage is in descriptor

Logic is used to create the Garage column.  If the observation contains the word 'Garage' it is copied into the new column  

In [164]:
data_merged['Parking'] = data_merged['Garage/Park'].apply(lambda x: x[1])

Retaining the format of the car spaces in the new Parking column.

In [165]:
data_merged['Garage'][data_merged['Garage']!= '0'].value_counts().sum()            ##### Matches above count

43135

In [166]:
data_merged['Garage'].value_counts()

2 Car     23083
1 Car     15135
0         14276
3 Car      3707
4 Car       702
5 Car       186
 Car        139
6 Car        95
7 Car        25
8 Car        24
9 Car        10
10 Car        9
12 Car        6
11 Car        5
13 Car        3
14 Car        2
18 Car        1
40 Car        1
50 Car        1
17 Car        1
Name: Garage, dtype: int64

In [167]:
len(data_merged[data_merged['Garage']==" Car"])

139

In [168]:
data_merged['Garage'] = data_merged['Garage'].apply(lambda x: '99999' if (x == ' Car') else x)         #####perhaps impute 99999 later

139 observations had responses that indicated garages but did not have a numerical value in the observation.  These are recoded as 99999 to maintain identity.

In [169]:
data_merged['Garage'].value_counts()

2 Car     23083
1 Car     15135
0         14276
3 Car      3707
4 Car       702
5 Car       186
99999       139
6 Car        95
7 Car        25
8 Car        24
9 Car        10
10 Car        9
12 Car        6
11 Car        5
13 Car        3
14 Car        2
40 Car        1
50 Car        1
18 Car        1
17 Car        1
Name: Garage, dtype: int64

In [170]:
data_merged['Garage'] = data_merged['Garage'].str.replace(' ','')  

In [171]:
data_merged['Garage'].head(10)

0    2Car
1    2Car
2       0
3       0
4    2Car
5    2Car
6    2Car
7    1Car
8    2Car
9    2Car
Name: Garage, dtype: object

In [172]:
data_merged['Garage'] = data_merged['Garage'].str.replace('Car','')  

In [173]:
data_merged['Garage'] = data_merged['Garage'].str.replace('car','')  

In [174]:
data_merged['Garage'] = pd.to_numeric(data_merged['Garage'])  

In [175]:
data_merged['Garage'].describe()

count    57411.000000
mean       243.463047
std       4914.473854
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max      99999.000000
Name: Garage, dtype: float64

In [176]:
data_merged = data_merged.drop('Garage/Park',axis=1)  #### dropp column

Garage/Parking is processed as follows:
1. The garage and parking values are split and stored in 2 new columns.
2. 2 incomplete Garage obervations were filled using a search on Zillow
3. All non-numeric characters are removed from the Garage. 
4. The series is converted to numeric values

-------------------------------------------------------------------------------

# Year Built

Year built is the year the building was built.  Age can be inferred

In [177]:
data_merged['Year Built'].describe()

count    57407.000000
mean      1964.288449
std         39.063627
min          0.000000
25%       1951.000000
50%       1969.000000
75%       1988.000000
max       2021.000000
Name: Year Built, dtype: float64

In [178]:
data_merged[data_merged['Year Built'] == 0]

#     MLS# Status Status Change Timestamp Prop Type           Address  \
1285  1  N353220   CLSD                12/22/20        SF  202 Pole Hill Rd   

         City  Acres  Sq Ft Total  Sq Ft Est Heated Above Grade  ...  \
1285  Bethany    1.5         2328                        2328.0  ...   

     Year Built    DOM       Listing Agent  List Office list_price  \
1285        0.0  1,945  Thomas J Cavaliere       CBRB34     369900   

     close_price Full Bath  Half Bath  Garage             Parking  
1285      400000         2          1       2  Under House Garage  

[1 rows x 23 columns]

In [179]:
data_merged.drop(index=1285,inplace = True)

In [180]:
data_merged[data_merged['Year Built'] <= 1650]

Empty DataFrame
Columns: [#, MLS#, Status, Status Change Timestamp, Prop Type, Address, City, Acres, Sq Ft Total, Sq Ft Est Heated Above Grade, Style or Rent Type, Rooms Total, Beds Total, Year Built, DOM, Listing Agent, List Office, list_price, close_price, Full Bath, Half Bath, Garage, Parking]
Index: []

[0 rows x 23 columns]

In [181]:
data_merged[data_merged['Year Built'] <= 1650]

Empty DataFrame
Columns: [#, MLS#, Status, Status Change Timestamp, Prop Type, Address, City, Acres, Sq Ft Total, Sq Ft Est Heated Above Grade, Style or Rent Type, Rooms Total, Beds Total, Year Built, DOM, Listing Agent, List Office, list_price, close_price, Full Bath, Half Bath, Garage, Parking]
Index: []

[0 rows x 23 columns]

In [182]:
data_merged[data_merged['Year Built'].isnull()]

#       MLS# Status Status Change Timestamp Prop Type  \
29413  1001  170392935   CLSD                06/18/21        SF   
52356   501  170348051   CLSD                12/29/20        SF   
54216  2501  170335485   CLSD                12/12/20        SF   
54560  2501  170319053   CLSD                01/25/21        SF   

                Address        City  Acres  Sq Ft Total  \
29413     25 Hanford Ln  New Canaan   2.31         3554   
52356   68 Bushnell Ave   Watertown   0.34         2083   
54216  24 10 O Clock Ln      Weston   2.12         2674   
54560      40 Bridge St    Westport   0.30         2331   

       Sq Ft Est Heated Above Grade  ... Year Built  DOM     Listing Agent  \
29413                        2984.0  ...        NaN   13  Christine D Saxe   
52356                        2083.0  ...        NaN   20  William Narducci   
54216                        1649.0  ...        NaN   38  Patricia D Manea   
54560                        1531.0  ...        NaN  110     Angus Beavers   

       List Office list_price close_price Full Bath  Half Bath  Garage  \
29413       CPCT83    1649000     1900000         3          0       2   
52356       SHWC22     287000      281750         1          1       2   
54216         4166     574995      585000         3          0       1   
54560         3548     975000      750000         1          1       1   

                                   Parking  
29413               Attached Garage, Paved  
52356  Detached Garage, Off Street Parking  
54216                      Detached Garage  
54560                      Attached Garage  

[4 rows x 23 columns]

In [183]:
### 52356 1920, 29413 1969, 54216 1956, 54560 2021

In [184]:
data_merged.at[52356,'Year Built'] = 1920

In [185]:
data_merged.at[29413,'Year Built'] = 1969

In [186]:
data_merged.at[54216,'Year Built'] = 1956

In [187]:
data_merged.at[54560,'Year Built'] = 2021

In [188]:
len(data_merged[data_merged['Year Built'].isnull()])

0

In [189]:
data_merged['Year Built'].describe()

count    57410.000000
mean      1964.322818
std         38.193826
min       1659.000000
25%       1951.000000
50%       1969.000000
75%       1988.000000
max       2021.000000
Name: Year Built, dtype: float64

### Age

In [190]:
today = pd.to_datetime('today')

In [191]:
current_year = today.year

In [192]:
data_merged['Age'] = current_year - data_merged['Year Built'] 

In [193]:
data_merged['Age'].describe()

count    57410.000000
mean        56.677182
std         38.193826
min          0.000000
25%         33.000000
50%         52.000000
75%         70.000000
max        362.000000
Name: Age, dtype: float64

____________________________________________________________________________

# Days on Market

Days on the market is number of days between when the day the property was listed and the day the property closed

In [194]:
data_merged['DOM'].head(20)

0      22
1      56
2       8
3     231
4      69
5       2
6      41
7      77
8       3
9      88
10      7
11     73
12      8
13      4
14    242
15      1
16      2
17     69
18      4
19      7
Name: DOM, dtype: object

In [195]:
data_merged['DOM'].describe()

count     57410
unique     1087
top           3
freq       2671
Name: DOM, dtype: object

The series needs to be converted to numeric values

In [196]:
print(data_merged['DOM'].value_counts())

3      2671
2      2229
4      2183
5      1557
1      1122
       ... 
587       1
926       1
570       1
873       1
-2        1
Name: DOM, Length: 1087, dtype: int64


In [197]:
data_merged['DOM'].isnull().value_counts()

False    57410
Name: DOM, dtype: int64

In [198]:
(data_merged['DOM']=='').value_counts()

False    57410
Name: DOM, dtype: int64

In [199]:
#########is numeric to look fin at this point

In [200]:
data_merged['DOM'] = data_merged['DOM'].apply(lambda x: str(x).replace(',','') if (',' in str(x)) else x )

In [201]:
data_merged['DOM'] = pd.to_numeric(data_merged['DOM']) 

Non numeric characters were removed and series converted to numeric values

In [202]:
data_merged['DOM'].isnull().value_counts()

False    57410
Name: DOM, dtype: int64

In [203]:
##data_merged[data_merged['DOM'].isnull()]

In [204]:
data_merged['DOM'].describe()

count    57410.000000
mean        52.470528
std         83.467637
min       -328.000000
25%          6.000000
50%         29.000000
75%         66.000000
max       2889.000000
Name: DOM, dtype: float64

In [205]:
len(data_merged[data_merged['DOM'] < 0])

11

Negative values do not make sense. These 11 observations will be replaced with the mean days on market

In [206]:
average_dom = (data_merged.loc[data_merged['DOM'] >= 0,'DOM']).mean()

In [207]:
data_merged.loc[data_merged['DOM'] < 0,'DOM'] = average_dom

In [208]:
data_merged['DOM'].describe()

count    57410.000000
mean        52.492064
std         83.443046
min          0.000000
25%          6.000000
50%         29.000000
75%         66.000000
max       2889.000000
Name: DOM, dtype: float64

Days on Market were processed as follows:
1. Non numeric characters were removed and the series was converted to numeric values
2. 11 negative observations were replaced by the mean Days on Market

***

# Summary Stats

In [209]:
print(data_merged.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 57410 entries, 0 to 58106
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   #                             57410 non-null  int64  
 1   MLS#                          57410 non-null  object 
 2   Status                        57410 non-null  object 
 3   Status Change Timestamp       57410 non-null  object 
 4   Prop Type                     57410 non-null  object 
 5   Address                       57410 non-null  object 
 6   City                          57410 non-null  object 
 7   Acres                         57410 non-null  float64
 8   Sq Ft Total                   57410 non-null  int64  
 9   Sq Ft Est Heated Above Grade  57410 non-null  float64
 10  Style or Rent Type            57410 non-null  object 
 11  Rooms Total                   57410 non-null  int64  
 12  Beds Total                    57410 non-null  int64  
 13  Y

In [210]:
print(data_merged.describe())

                  #         Acres   Sq Ft Total  Sq Ft Est Heated Above Grade  \
count  57410.000000  57410.000000  57410.000000                  57410.000000   
mean    1686.908378      1.609958   2149.181780                   1943.566086   
std     1466.587461     80.129240   1329.153322                   1169.310715   
min        1.000000      0.000000    100.000000                    100.000000   
25%        1.000000      0.130000   1302.000000                   1210.000000   
50%     1501.000000      0.340000   1794.000000                   1592.000000   
75%     3001.000000      1.000000   2565.000000                   2323.750000   
max     4501.000000  12911.000000  61184.000000                  61184.000000   

        Rooms Total    Beds Total    Year Built           DOM    list_price  \
count  57410.000000  57410.000000  57410.000000  57410.000000  5.741000e+04   
mean       6.883313      3.158091   1964.322818     52.492064  4.528182e+05   
std        2.296270      1.021816

In [211]:
data_merged.reset_index(inplace=True,drop = True)